server/commands/write_authzmodel.go

```diff
@@ -163,3 +185,83 @@ func isUsersetRewriteValid(allRelations map[string]struct{}, relationsOnType map
 
 	return nil
 }
+
+func areRelationalTypesValid(tds []*openfgapb.TypeDefinition) error {
+	typeToRelations := map[string]map[string]struct{}{}
+	for _, td := range tds {
+		typeName := td.GetType()
+		typeToRelations[typeName] = map[string]struct{}{}
+		for relationName := range td.GetRelations() {
+			typeToRelations[typeName][relationName] = struct{}{}
+		}
+	}
+
+	// Here we are checking that every type on a relation is valid. This means:
+	// 1. If it is a type (e.g. user) then this type is a type in the model
+	// 2. If it is a type#relation then this type with this relation is in the model
+	for _, td := range tds {
+		for _relation, _ := range td.GetRelations() {
+			metadata, ok := td.GetMetadata().GetRelations()[_relation]
```

Comment:
if people give us some relations with types defined but others with nothing (i.e. `td.GetMetadata() == nil`), we should throw an error

Reply:
Yes. In this case, if so, then `ok` should be false.

Reply:
If `ok` is false this indicates the relation had no type information associated with it, correct?

e.g.

```
type user

type document
  relations
    define reader: [user]
    define can_read: reader
```

Is that correct?

Reply:
Correct. Or for some reason the JSON doesn't contain it.

Reply:
If it doesn't contain it then we assume it is not assignable, and if it references `self` then the logic below (determining it's assignability) would fail yeah?